src/log.c

```diff
@@ -36,70 +36,297 @@
 #include <stdarg.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
+#include <inttypes.h>
 
+#include <rdma/fi_errno.h>
 #include "fi.h"
 #include "fi_log.h"
 
+#define ARRAY_SIZE(x) (sizeof((x)) / sizeof((x)[0]))
+
+static struct log_providers provider_list;
+
+static const char *subsystems[] = {
+	[FABRIC] = "fabric",
+	[DOMAIN] = "domain",
+	[CM] = "cm",
+	[EP] = "ep",
+	[AV] = "av",
+	[CQ] = "cq",
+	[EQ] = "eq",
+	[MR] = "mr"
+};
+
+static const char *log_levels[] = {
+	[LOG_WARN] = "warn",
+	[LOG_TRACE] = "trace",
+	[LOG_INFO] = "info"
+};
+
+/*
+ * By default support logging from all providers and subsystems. This will
+ * change if the relevant environment variables are set.
+ */
+uint64_t log_mask = PROV_MASK | SUBSYS_MASK;
+
+typedef void found_func_t(char *ptr);
+
 /* General implementation note: these functions currently use multiple fprintfs
  * in a row, which can render in an ugly fashion for multithreaded code and for
  * some mpirun implementations.  If this bugs anyone enough then we can convert
  * them to snprintf to build up the printout in a single buffer.
```

Comment:
Is this comment still true?

If so, it would be a nice target of opprotunity to fix this now, while mucking around in the rest of the logging stuff (because, as noted, multiple printfs at likely to be broken up in practice by parallel run-time environments such as those commonly used by mpirun implementations).